### PR TITLE
[PaddleInference] support ptq and cachekv_quant in BlockMultiHeadAttention op

### PR DESCRIPTION
**Pull request by RichardWooSJTU**
### PR types

Function optimization
### PR changes

OPs
### Description

For block attention
1. support dynamic cachekv quant
2. support static cachekv quant
3. support PTQ fusion

Pcard-71502

### CODE REVIEW

File: paddle/phi/kernels/gpu/c_embedding_kernel.cu [hunk 1]
```diff
@@ -41,16 +41,15 @@ __global__ void CEmbedding(T* out,
...
 
-    PADDLE_ENFORCE(
-        id >= 0 && (vocab_size < 0 || id < vocab_size),
-        "The index is out of bounds, "
-        "please check whether the dimensions of index and "
-        "input meet the requirements. It should "
-        "be less than [%d] and greater than or equal to 0, but received [%d]",
-        vocab_size,
-        id);
+    // PADDLE_ENFORCE(
```
Comment (zhoutianzi666) [inline paddle/phi/kernels/gpu/c_embedding_kernel.cu]: 这个为啥注释掉？

Comment (MARD1NO) [inline paddle/phi/kernels/gpu/c_embedding_kernel.cu]: 这个检查还是要的

**Review by MARD1NO — APPROVED** [review]

File: paddle/phi/kernels/gpu/c_embedding_kernel.cu (hunk 1 above)
Comment (RichardWooSJTU) [reply paddle/phi/kernels/gpu/c_embedding_kernel.cu]: 目前多轮会报错在这里 报错不符合预期 还在找作者debug

Comment (RichardWooSJTU) [reply paddle/phi/kernels/gpu/c_embedding_kernel.cu]: 目前多轮会报错在这里 报错不符合预期 还在找作者debug

**Review by Aurelius84 — APPROVED** [review]

**Review by vivienfanghuagood — APPROVED** [review]
LGTM for API change

**Review by heavengate — APPROVED** [review]

File: python/paddle/incubate/nn/functional/block_multihead_attention.py [hunk 2]
```diff
@@ -30,12 +30,26 @@ def block_multihead_attention(
...
     pre_key_cache=None,
     pre_value_cache=None,
+    cache_k_quant_scales=None,
+    cache_v_quant_scales=None,
+    cache_k_dequant_scales=None,
+    cache_v_dequant_scales=None,
+    qkv_out_scale=None,
+    qkv_bias=None,
+    out_shift=None,
+    out_smooth=None,
```
Comment (jeff41404) [inline python/paddle/incubate/nn/functional/block_multihead_attention.py:40]: The newly added parameters, such as `cache_k_quant_scales`, need to be described in docstring of section `Args` below

Comment (jeff41404) [reply python/paddle/incubate/nn/functional/block_multihead_attention.py:40]: Normally, the added parameters should be added at the end to ensure compatibility

Comment (RichardWooSJTU) [reply python/paddle/incubate/nn/functional/block_multihead_attention.py:40]: This will be added in next PR.

**Review by jeff41404 — APPROVED** [review]

**Review by Ligoml — APPROVED** [review]
LGTM for docs（需要补充中文文档）

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**